### PR TITLE
Feature: RxJava2 Support

### PR DESCRIPTION
**Pull request by cmelchior**
Part of #4785 

This feature branch contains all changes required to convert our current RxJava1 support to RxJava2.

There has been a lot of discussion about how our RxJava support could look like in #3497 and #4291 . Mostly it came down to a choice between splitting into a separate library like `RxRealm` or continue to use our "optional dependency" approach. 

After thinking about this a lot and trying to prototype the various approaches, I came to the conclusion that we should continue to use our "optional dependency" approach.

Reasons:

* Splitting RxJava out is mostly a benefit for our codebase, not so much for users of Realm.
* Having a fluent API is extremely nice: `realm.where(Person.class).findAll().asFlowable()` vs. `RxRealm.asFlowable(realm.where(Person.class).findAll())`.
* With Kotlin we could introduce the fluent API as extension functions, but right now the numbers still only show ~15% of Realm uses having Kotlin in their code base.
* People not using RxJava are not really affected either way. We had a few cases where reflection based libraries broke, like Jackson, but apart from that, we had no complaints about the optional dependency approach.
* RxJava1 support is still possible through a `RxRealm` abstraction
* We don't have to spend time setting up a new repo, CI, documentation, and distribution for a new library.

@realm/java Let me know if you disagree strongly with this.

TODO:
- [x] [In Review] Convert existing interfaces and method to RxJava2. Mostly Observable -> Flowable.
- [x] [In Review] Implement `RealmResults.asChangesetObservable()`, `RealmList.asChangesetObservable()` and `RealmObject.asChangesetObservable()`. Wil require moving `Pair` to public API.
- [ ] [POSTPONED] Implement `Realm.asCompletable()` and `DynamicRealm.asCompletable()` for transactions.
- [x] Consider if `RxObservableFactory` is still the correct name. `Observable` has a specific meaning in the context of RxJava 2. `RxStreamFactory` perhaps? -> Not immediately clear that changing this name is a massive improvement, so no reason to break it for a lot of people.
- [x] Update example app to RxJava2 ( @dalinaum)
- [x] Go through wording and Javadoc to ensure it matches RxJava2

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,21 @@
+## 4.0.0 (YYYY-MM-DD)
+
+### Breaking Changes
+
+* Realm has upgraded its RxJava1 support to RxJava2 (#3497)
+  * `Realm.asObservable()` has been renamed to `Realm.asFlowable()`.
+  * `RealmList.asObservable()` has been renamed to `Realm.asFlowable()`.
```
Comment (Zhuinden) [inline CHANGELOG.md]: `Realm.`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,21 @@
...
+  * `RealmResults.asObservable()` has been renamed to `Realm.asFlowable()`.
```
Comment (Zhuinden) [inline CHANGELOG.md]: `Realm.`?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,21 @@
...
+  * `RealmObject.asObservable()` has been renamed to `Realm.asFlowable()`.
```
Comment (Zhuinden) [inline CHANGELOG.md]: `Realm.`?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,21 @@
...
+
+### Enhancements
+
+  * `RealmResults.asObservable()` has been added that publish a `Pair<RealmResults, OrderedCollectionChangeSet>` (#4277).
+  * `RealmList.asObservable()` has been added that publish a `Pair<RealmList, OrderedCollectionChangeSet>` (#4277).
```
Comment (Zhuinden) [inline CHANGELOG.md]: Is Pair from `android.util` or your own class?

**Comment by Zhuinden** [comment]
I assume `asFlowable` with `LATEST` strategy, right?

File: CHANGELOG.md (hunk 4 above)
Comment (cmelchior) [inline CHANGELOG.md]: Our own. We cannot depend on `android.util.Pair` on the JVM, sadly :(

**Comment by cmelchior** [comment]
Yes, it would be `Flowable` with `LATEST` ... changeset observers would have to be `Observable()` though as you will need all `OrderedCollectionChangeset` events in order to animate things correctly.

**Comment by Zhuinden** [comment]
@curiousily you can query Realm for changes only on looper threads, which means the UI thread, and other handler threads.

So it's the same as now except with `Flowable` and `Observable` where applicable, with additional suport for `changeSets` as well (fine-grained notifications - `OrderedRealmCollectionChangeListener` and `RealmObjectChangeListener`)

**Pull request merged**